zarr/core.py

```diff
@@ -1572,6 +1572,12 @@ def _chunk_getitem(self, chunk_coords, chunk_selection, out, out_selection,
 
         assert len(chunk_coords) == len(self._cdata_shape)
 
+        out_is_ndarray = True
+        try:
+            out = ensure_ndarray(out)
```

Comment:
Note that this is already checked by `test_get_selection_out`. Without this `try`/`except`, we get the following `TypeError` because `out` is a Zarr `Array`, which cannot be coerced to an `ndarray`. This is ok and intentional. So we just carry on without coercing that case and note that we do not have an `ndarray` when checking later.

<details>

```python
_________________________ test_get_selection_out ____________________________

    def test_get_selection_out():
    
        # basic selections
        a = np.arange(1050)
        z = zarr.create(shape=1050, chunks=100, dtype=a.dtype)
        z[:] = a
        selections = [
            slice(50, 150),
            slice(0, 1050),
            slice(1, 2),
        ]
        for selection in selections:
            expect = a[selection]
            out = zarr.create(shape=expect.shape, chunks=10, dtype=expect.dtype, fill_value=0)
>           z.get_basic_selection(selection, out=out)

zarr/tests/test_indexing.py:1036: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
zarr/core.py:698: in get_basic_selection
    fields=fields)
zarr/core.py:740: in _get_basic_selection_nd
    return self._get_selection(indexer=indexer, out=out, fields=fields)
zarr/core.py:1028: in _get_selection
    drop_axes=indexer.drop_axes, fields=fields)
zarr/core.py:1573: in _chunk_getitem
    out = ensure_ndarray(out)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

buf = <zarr.core.Array (100,) int64>

    def ensure_ndarray(buf):
        """Convenience function to coerce `buf` to a numpy array, if it is not already a
        numpy array.
    
        Parameters
        ----------
        buf : array-like or bytes-like
            A numpy array or any object exporting a buffer interface.
    
        Returns
        -------
        arr : ndarray
            A numpy array, sharing memory with `buf`.
    
        Notes
        -----
        This function will not create a copy under any circumstances, it is guaranteed to
        return a view on memory exported by `buf`.
    
        """
    
        if isinstance(buf, np.ndarray):
            # already a numpy array
            arr = buf
    
        elif isinstance(buf, array.array) and buf.typecode in 'cu':
            # Guard condition, do not support array.array with unicode type, this is
            # problematic because numpy does not support it on all platforms. Also do not
            # support char as it was removed in Python 3.
            raise TypeError('array.array with char or unicode type is not supported')
    
        else:
    
            # N.B., first take a memoryview to make sure that we subsequently create a
            # numpy array from a memory buffer with no copy
    
            if PY2:  # pragma: py3 no cover
                try:
                    mem = memoryview(buf)
                except TypeError:
                    # on PY2 also check if object exports old-style buffer interface
                    mem = np.getbuffer(buf)
    
            else:  # pragma: py2 no cover
>               mem = memoryview(buf)
E               TypeError: memoryview: a bytes-like object is required, not 'Array'

.tox/py36/lib/python3.6/site-packages/numcodecs/compat.py:74: TypeError
```
</details>

ref: https://travis-ci.org/zarr-developers/zarr-python/jobs/610533967#L2316